src/ripple/basics/SlabAllocator.h

```diff
@@ -55,42 +55,80 @@ class SlabAllocator
         std::mutex m_;
 
         // A linked list of appropriately sized free buffers:
-        std::uint8_t* l_ = nullptr;
+        std::uint8_t* list_ = nullptr;
 
-        // The next memory block
+        // The next block:
         SlabBlock* next_;
 
         // The underlying memory block:
-        std::uint8_t const* const p_ = nullptr;
+        std::uint8_t* data_ = nullptr;
 
-        // The extent of the underlying memory block:
-        std::size_t const size_;
+        // The number of items available in this block
+        std::size_t avail_{0};
 
-        SlabBlock(
-            SlabBlock* next,
+        explicit SlabBlock(SlabBlock* next) : next_(next)
+        {
+        }
+
+        // We don't release the allocated memory here: slabs are only
+        // destroyed when the process is shutting down, and we cannot
```

Comment:
No, not quite: what this comment says is that the *destructor* of a `SlabBlock` that points to an allocated chunk of memory does not deallocate that chunk. Of course, this destructor will _only_ ever be called when the process is shutting down, at which point any and all allocated memory will be released back to the O/S.

Note that when a `SlabBlock` becomes empty, the buffer it points to is dynamically deallocated. But as the code is written, during shutdown it's possible for the destructor of a SlabBlock to be called while items are still allocated out of the slab block.

It's possible to fix this but I don't know if the computational overhead is worth it for a hot path.

Reply:
Okay, the dynamic deallocation of unused buffers occurs through `boost::alignment::aligned_free(ptr);` calls, am I right?

Why is the fix complicated? If we can ensure that the destructor of `SlabBlock` is made after all the other objects have been destroyed, we will not face this issue. 

I'm guessing the shutdown process is concurrent with multiple threads, hence causes this stated overhead?

Reply:
It’s complicated because the order of destruction of objects isn’t, necessarily, trivial or “fixed” and depends of a lot of variables that aren’t always directly controllable and, to the degree that they are, to exercise control can be onerous.

It’s unlikely that a program of even moderate complexity could successfully manage the ordering without relying on compiler extensions, and do so reliably in the face of changes in the codebase and the minutiae of compiling.

As I explained, when the process shuts down, the O/S reclaims all the allocated memory, whether it was freed or not. IMO it seems counterproductive to invest too much time and effort and to complicate the code in order to ensure we release the memory block a few milliseconds before the process shuts down.

Reply:
alright, that's fair. thanks for the explanation!